orbis-kernel/src/sys/sys_environment.cpp

```diff
@@ -1,7 +1,85 @@
 #include "sys/sysproto.hpp"
 
+#include "KernelContext.hpp"
+
 orbis::SysResult orbis::sys_kenv(Thread *thread, sint what,
                                  ptr<const char> name, ptr<char> value,
                                  sint len) {
-  return ErrorCode::NOSYS;
+  enum action { kenv_get, kenv_set, kenv_unset, kenv_dump };
+
+  auto &[kenv, _] = thread->tproc->context->getKernelEnv();
+
+  if (what == kenv_dump) {
+    int needed = 0;
+    int done = 0;
+
+    for (const auto &[key, env_value] : kenv) {
+      size_t entry = 0;
+      // Entry: size of both full buffers, the '=' and the '\0' at the end
+      if (value == nullptr || len == 0) {
+        entry = key.size() + 1 + strnlen(env_value, 128) + 1;
+      } else {
+        char buf[128 * 2 + 2];
+
+        char *_buf = buf;
+        std::strncpy(buf, key.data(), key.size());
+        _buf += key.size();
+
+        *_buf++ = '=';
+
+        const size_t value_size = strnlen(env_value, 128);
+        std::strncpy(_buf, env_value, value_size);
+        _buf += value_size;
+
+        *_buf++ = '\0';
+
+        entry = _buf - buf;
+        ORBIS_RET_ON_ERROR(uwriteRaw(value + done, buf, entry));
+        len -= entry;
+        done += entry;
+      }
+      needed += entry;
+    }
+
+    if (done == needed) {
+      return {};
+    }
+    return {static_cast<ErrorCode>(-needed)};
```

Comment:
To return value use thread->retval[0]
Non zero error code will report about error to thr user space

Reply:
Fixed